### PR TITLE
Add support for mods and vips parsing

### LINKED ISSUE #153: Add neater hooks for the mods and vips commands.
Add a function that returns the responses from `/vips` and `/mods`

### PR DESCRIPTION
**Pull request by TroyKomodo**
fixes #153

### CODE REVIEW

**Comment by TroyDota** [comment]
I tried to make it concurrent safe. likely crappy implementation of that.

File: client.go [hunk 1]
```diff
@@ -595,6 +609,96 @@ func (c *Client) FollowersOff(channel string) {
 	c.Say(channel, "/followersoff")
 }
 
+// GetVips run twitch command `/vips` with the given channel in argument and returns a string slice of all vips
+func (c *Client) GetVips(channel string) []string {
+	channel = strings.ToLower(channel)
+	ch := make(chan NoticeMessage)
+	makeNew := true
+
+	defer func() {
```
Comment (TroyDota) [inline client.go:618]: i wonder if its possible to hit a race condition on delete after a new one has been created. if the function is called at a very specific time.

**Comment by TroyDota** [comment]
I don't really like it much to be completely fair. I feel like there is probably a better way to make it goroutine safe. It just looks messy.

**Comment by TroyDota** [comment]
```
Feb 15 01:48:53.965 [INFO] (twitch.NoticeMessage) {
 Raw: (string) (len=135) "@msg-id=vips_success :tmi.twitch.tv NOTICE #troydota :The VIPs of this channel are: admiralbulldog, aetaric, ales_, eeddya, komodotroy.",
 Type: (twitch.MessageType) 6,
 RawType: (string) (len=6) "NOTICE",
 Tags: (map[string]string) (len=1) {
  (string) (len=6) "msg-id": (string) (len=12) "vips_success"
 },
 Message: (string) (len=81) "The VIPs of this channel are: admiralbulldog, aetaric, ales_, eeddya, komodotroy.",
 Channel: (string) (len=8) "troydota",
 MsgID: (string) (len=12) "vips_success"
}

Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
Feb 15 01:48:53.965 [INFO] [admiralbulldog aetaric ales_ eeddya komodotroy]
```

```go
client.OnNoticeMessage(func(message twitch.NoticeMessage) {
	log.Infoln(spew.Sdump(message))
	if message.MsgID == "vips_success" {
		client.vipsMtx.RLock()
		if v, ok := client.vipsChan[message.Channel]; ok {
			client.vipsMtx.RUnlock()
			v.mutex.Lock()
			v.finished = true
			for _, ch := range v.chans {
				ch <- message
			}
			v.mutex.Unlock()
		} else {
			client.vipsMtx.RUnlock()
		}
	} else if message.MsgID == "room_mods" {
		client.modsMtx.RLock()
		if v, ok := client.modsChan[message.Channel]; ok {
			client.modsMtx.RUnlock()
			v.mutex.Lock()
			v.finished = true
			for _, ch := range v.chans {
				ch <- message
			}
			v.mutex.Unlock()
		} else {
			client.modsMtx.RUnlock()
		}
	}
})

client..OnConnect(func() {
	for v := 0; v < 10; v++ {
		go func() {
			log.Println(client.GetVips("troydota"))
		}()
	}
})
```

so it seems that the concurrency works perfectly. Since we only send the `/vips` command once but all goroutines get a response.

**Comment by RAnders00** [comment]
As far as I can see this is missing a timeout currently- You definitely want that I think. Error cases are also for example a connection disconnect, which could be handled too.

**Comment by 3ventic** [comment]
The message format isn't guaranteed to be static for either of the responses. I would recommend grabbing the string between `:`, and end of string or `.`, whichever comes first. This should also allow combining a lot of the common logic between the two even without generics.

**Comment by TroyDota** [comment]
https://play.golang.org/p/2LNyI4GbQuh

**Comment by TroyDota** [comment]
Also, I know I should write tests for this, but to be completely fair. I am not entirely sure how to write an effective test.  That will test, both parsing and race conditions.

**Review by gempir — CHANGES_REQUESTED** [review]
Thanks for the contribution. It's certainly a cool and useful feature. 

I don't expect everything to be fully tested, but I do expect some level of tests. 

Especially parsing should be well tested. Keep edge cases in mind like when a channel has no VIPs or Mods all that can be nicely tested with a table driven test.

https://github.com/golang/go/wiki/TableDrivenTests

File: client_test.go [hunk 2]
```diff
@@ -2161,3 +2161,215 @@ func TestEmptyCapabilities(t *testing.T) {
...
+		if strings.HasPrefix(message, "PRIVMSG") {
+			receivedMsg = message
+			// the fact we close this and it doesnt panic means that only one message is received.
+			close(ch)
+		}
+	})
+
+	client := newTestClient(host)
+
+	go client.Connect()
```
Comment (TroyDota) [inline client_test.go:2180]: why did this fail when i literally copied it from the test above?

File: client_test.go [hunk 3]
```diff
@@ -2161,3 +2161,215 @@ func TestEmptyCapabilities(t *testing.T) {
...
+	}
+
+	// wait for channel resolve before waiting on wait group because we then know the server doesnt receive the message twice.
+	<-ch
+	wg.Wait()
+
+	assertStringsEqual(t, "PRIVMSG #gempir :/mods", receivedMsg)
+}
+
+func TestVipsModsParsing(t *testing.T) {
```
Comment (TroyDota) [inline client_test.go:2219]: uhm, do i just remove some vips?

File: client_test.go (hunk 2 above)
Comment (pajlada) [reply client_test.go:2180]: The linting is on a "lint new things, but not old things" since it was implemented. See other tests, e.g. `connectAndEnsureGoodDisconnect` (which you might want to use anyway)

**Comment by gempir** [comment]
I updated the parsing tests. Would like to see some more with like 1 mod/vip. Also maybe some like just a random string and testing what happens.

I also for sure want the other client code tested not only the parsing, not every possible case but some basic tests for sure.

Also make sure to always compile with -race to better see any race conditions you build. https://golang.org/doc/articles/race_detector

**Comment by gempir** [comment]
Great  start for a test, if possible please try also writing fake values from the server so the client starts reading and parsing them. 

Also the linter has some issues with your test, maybe split it up a bit or something.

**Comment by TroyDota** [comment]
Sure

**Comment by gempir** [comment]
Also I tried it out and I'm just getting timeouts? 
```
	client.OnPrivateMessage(func(message twitch.PrivateMessage) {
		fmt.Println(message.Message)

		mods, err1 := client.GetMods("gempir", time.Second*5)
		if err1 != nil {
			log.Println(err1)
		}

		log.Println(mods)

		vips, err2 := client.GetVips("gempir", time.Second*5)
		if err2 != nil {
			fmt.Println(err2)
		}

		log.Println(vips)
	})

	client.Join("gempir")

	err := client.Connect()
	if err != nil {
		panic(err)
	}
```


```
2021/02/25 20:50:44 request timedout
2021/02/25 20:50:44 []
request timedout
2021/02/25 20:50:49 []
```

File: client.go [hunk 4]
```diff
@@ -595,6 +612,159 @@ func (c *Client) FollowersOff(channel string) {
...
+			case msg = <-ch:
+			}
+		}
+	}
+
+	return parseVipsOrModsMsg(msg.Message), nil
+}
+
+// GetMods run twitch command `/mods` with the given channel in argument and returns a string slice of all mods
+func (c *Client) GetMods(channel string, timeout time.Duration) ([]string, error) {
```
Comment (gempir) [inline client.go:699]: This should be an option on the client instead that you can configure, this would be too confusing for an end-user to decide which timeout is approriate.

File: client.go [hunk 5]
```diff
@@ -41,6 +41,9 @@ var (
 	// ErrConnectionIsNotOpen is returned by Disconnect in case you call it without being connected
 	ErrConnectionIsNotOpen = errors.New("connection is not open")
 
+	// ErrRequestTimedout is returned by GetVips or GetMods if a response is not returned after your specified timeout
+	ErrRequestTimedout = errors.New("request timedout")
```
Comment (gempir) [inline client.go:45]: timed out not timedout

**Comment by TroyDota** [comment]
Strange. Sorry ive been a bit busy with another project and do intend to fix and finish this pr. I should finish it this week.

**Comment by gempir** [comment]
There is no rush for me. Take your time

**Comment by gempir** [comment]
As per https://discuss.dev.twitch.tv/t/deprecation-of-chat-commands-through-irc/40486 

These commands are deprecated. It wouldn't make sense adding them.
